packages/medusa/src/services/discount.ts

```diff
@@ -536,6 +559,57 @@ class DiscountService extends BaseService {
       return Promise.resolve()
     })
   }
+
+  async upsertDiscountCondition_(
+    discountId: string,
+    data: UpsertDiscountConditionInput
+  ): Promise<void> {
+    const res = this.atomicPhase_(
+      async (manager) => {
+        const discountConditionRepo: DiscountConditionRepository =
+          manager.getCustomRepository(this.discountConditionRepository_)
+
+        if (data.id) {
+          return await discountConditionRepo.addConditionResources(
+            data.id,
+            data.resource_ids,
+            data.resource_type,
+            true
+          )
+        }
+
+        const discount = await this.retrieve(discountId, {
+          relations: ["rule", "rule.conditions"],
+        })
+
+        const created = discountConditionRepo.create({
+          discount_rule_id: discount.rule_id,
+          operator: data.operator,
+          type: data.resource_type,
+        })
+
+        const discountCondition = await discountConditionRepo.save(created)
+
+        return await discountConditionRepo.addConditionResources(
+          discountCondition.id,
+          data.resource_ids,
+          data.resource_type
+        )
+      },
+      async (err: any) => {
+        if (err.code === "23505") {
+          // A unique key constraint failed meaning the combination of
+          // discount rule id, type, and operator already exists in the db.
+          throw new MedusaError(
+            MedusaError.Types.DUPLICATE_ERROR,
+            `Discount Condition with operator '${data.operator}' and type '${data.resource_type}' already exist on a Discount Rule`
+          )
+        }
```

Comment:
Nit: Not sure if it works in this case but maybe consider using @pKorsholm's `formatException`?